### PR TITLE
Exclude weekends when considering bake time for promotion to next tenant

### PR DESCRIPTION
**Pull request by sahanocto**
# Background
This PR excludes weekends when determining if sufficient bake time has passed for a deployment to be eligible for promotion to the next tenant.

https://trello.com/c/fbVcavCR/349-cautious-promoter-should-only-count-business-days

## Before
A deployment that completed on a Friday that requires a bake time of two days before being promoted to the next tenant, would be promoted first thing Monday morning (which is the next earliest trigger of the scheduled runbook after it finished baking on the weekend).

## After
The same deployment would now be promoted on Tuesday. The two days of bake time is composed of the remaining hours left in the day on Friday + whole day Monday + the final remainder on Tuesday.

# Details

There is a long outstanding enhancement to NodaTime to distinguish between weekdays and weekends: https://github.com/nodatime/nodatime/issues/6
In the absence of that, given that the Cautious Deployer runbook is only triggered during weekdays, an approximation can be used to compensate for the additional wall-clock time needed to reach the desired bake time. That is, by checking at every one day interval since the last deployment, if this time falls on weekend, the "time since last deployment" calculation is adjusted by one day at a time.

Only the TenantPromotionStrategy has been updated. The (old?) EnvironmentPromotionStrategy has been left alone but can have the same change applied if needed.

### CODE REVIEW

File: source/Cli/Support/ClockExtensions.cs [hunk 1]
```diff
@@ -12,5 +13,22 @@ public static bool ItIsOutsidePromotionHours(this ZonedClock clock)
...
             return weekendDays.Contains(dow) || currentDateTime.TimeOfDay > new LocalTime(14, 0) || currentDateTime.TimeOfDay < new LocalTime(8, 0);
         }
+
+        public static TimeSpan WeekdayTimeSince(this IClock clock, DateTimeOffset since)
+        {
+            var currentDateTime = clock.GetCurrentInstant().ToDateTimeOffset();
+            var weekendCompensation = TimeSpan.Zero;
+            var weekendDays = new[] {DayOfWeek.Saturday, DayOfWeek.Sunday};
+
+            for (var date = since; date < currentDateTime; date += TimeSpan.FromDays(1))
```
Comment (sahanocto) [inline source/Cli/Support/ClockExtensions.cs:23]: Used a simple for loop here for the calculation. There's more complicated ways to do this with a bunch of date arithmetic.
- Assuming the Cautious Promoter is never executed on a weekend the result will be accurate (otherwise you could end up with negative TimeSpans being returned with the current logic, which won't affect the behaviour of the Cautious Promoter but the negative values will appear in the logging output). Given that the intention is for things to never finish baking on a weekend (as per the runbook schedule), I wasn't planning on fixing this.
- Assuming the bake times will be small (i.e. in the order of days not years) there won't be any performance problem.

File: source/Cli.Tests/ClockExtensionsFixture.cs [hunk 2]
```diff
@@ -26,6 +26,19 @@ public void OutsidePromotionHours(IsoDayOfWeek dow, int hour, bool expectedResul
...
         }
 
-
+        [TestCase(0, 0)]
+        [TestCase(24, 0)]
+        [TestCase(48, 0)]
+        [TestCase(55, 7)]
+        [TestCase(72, 24)]
+        [TestCase(100, 52)]
+        public void WeekdayTimeSince(int hours, int expectedResult)
```
[the remark this answers is not in the document]
Comment (benPearce1) [reply source/Cli.Tests/ClockExtensionsFixture.cs:35]: Nitpick: 
```suggestion
        public void WeekdayTimeSince(int hours, int expectedResultHours)
```

File: source/Cli.Tests/ClockExtensionsFixture.cs [hunk 3]
```diff
@@ -26,6 +26,19 @@ public void OutsidePromotionHours(IsoDayOfWeek dow, int hour, bool expectedResul
...
+        {
+            var initial = Instant.FromUtc(2021, 7, 26, 7, 0);
```
Comment (benPearce1) [inline source/Cli.Tests/ClockExtensionsFixture.cs:37]: Can we get a comment here this is a Monday morning

File: source/Cli/Strategies/TenantPromotionStrategy.cs [hunk 4]
```diff
@@ -295,7 +295,7 @@ bool BakeTimeHasElapsed(TenantConfiguration tenantConfiguration, DashboardItemRe
 
         private TimeSpan TimeSinceLastDeployment(DashboardItemResource deploymentResource)
         {
-            var timeSinceLastDeployment = clock.GetCurrentInstant().ToDateTimeOffset() - deploymentResource.CompletedTime.Value;
+            var timeSinceLastDeployment = clock.WeekdayTimeSince(deploymentResource.CompletedTime.Value);
```
Comment (benPearce1) [inline source/Cli/Strategies/TenantPromotionStrategy.cs:298]: Scope creep: can we add including weekdays as a configuration option, similar to the `DryRun` flag?

**Review by benPearce1 — APPROVED** [review]
Some comments around the test, but otherwise looks good.
If it is not too hard would it be possible to try and fix the calculation if the current time is on a weekend?

File: source/Cli/Support/ClockExtensions.cs (hunk 1 above)
Comment (flin-8) [reply source/Cli/Support/ClockExtensions.cs:23]: Let's make sure the code works no matter when it runs - there should be an easy way to get it to return zero on a weekend

File: source/Cli.Tests/ClockExtensionsFixture.cs (hunk 3 above)
Comment (flin-8) [reply source/Cli.Tests/ClockExtensionsFixture.cs:37]: Just rename the variable to `mondayMorning` etc.

File: source/Cli/Strategies/TenantPromotionStrategy.cs [hunk 5]
```diff
@@ -295,7 +295,7 @@ bool BakeTimeHasElapsed(TenantConfiguration tenantConfiguration, DashboardItemRe
...
             return timeSinceLastDeployment;
         }
```
Comment (flin-8) [inline source/Cli/Strategies/TenantPromotionStrategy.cs:300]: Discussing as a code comment coz it's easier:

@benPearce1 do you think we should just delete the environment promotion strategy? otherwise it's extra maintenance/drift

Comment (benPearce1) [reply source/Cli/Strategies/TenantPromotionStrategy.cs:300]: Yes we should.

File: source/Cli/Strategies/TenantPromotionStrategy.cs (hunk 4 above)
Comment (sahanocto) [reply source/Cli/Strategies/TenantPromotionStrategy.cs]: Done! `ExcludeWeekends` option added to `appsettings.json` and defaulted it to `true`

File: source/Cli.Tests/TenantPromotionStrategyFixture.cs [hunk 6]
```diff
@@ -375,6 +399,29 @@ public async Task SingleRelease_DeploymentCompleted_StillBaking_ShouldDoNothing(
...
+        public async Task SingleRelease_SingleDeployment_StillBakingDueToWeekend_ShouldDoNothing()
+        {
+            progressionResourceResult = new ProgressionResource
+            {
+                ChannelEnvironments = ChannelEnvironmentsReferenceData,
+                Environments = EnvironmentReferenceData,
+                Releases = new List<ReleaseProgressionResource>
+                {
+                    CreateRelease("0.0.1", ReferenceCollection.One(ProdEnvironment.Id))
+                        .WithDeployment(6.DaysAgo(clock), FirstTestTenant, TestEnvironment, DefaultChannel)
```
[the remark this answers is not in the document]
Comment (sahanocto) [reply source/Cli.Tests/TenantPromotionStrategyFixture.cs]: Clock is now explicitly reset for these tests

**Comment by sahanocto** [comment]
[Cautious Promoter should only count business days](https://trello.com/c/fbVcavCR/349-cautious-promoter-should-only-count-business-days)

File: source/Cli/Strategies/TenantPromotionStrategy.cs [hunk 7]
```diff
@@ -295,8 +295,13 @@ bool BakeTimeHasElapsed(TenantConfiguration tenantConfiguration, DashboardItemRe
 
         private TimeSpan TimeSinceLastDeployment(DashboardItemResource deploymentResource)
         {
-            var timeSinceLastDeployment = clock.GetCurrentInstant().ToDateTimeOffset() - deploymentResource.CompletedTime.Value;
-            return timeSinceLastDeployment;
+            if (configuration.GetValue<bool>("ExcludeWeekends"))
```
Comment (flin-8) [inline source/Cli/Strategies/TenantPromotionStrategy.cs:298]: Extract `ExcludeWeekends` to a constant, and we can reuse in the tests

File: source/Cli/Support/ClockExtensions.cs [hunk 8]
```diff
@@ -12,5 +13,40 @@ public static bool ItIsOutsidePromotionHours(this ZonedClock clock)
...
             return weekendDays.Contains(dow) || currentDateTime.TimeOfDay > new LocalTime(14, 0) || currentDateTime.TimeOfDay < new LocalTime(8, 0);
         }
+
+        public static TimeSpan WeekdayTimeSince(this IClock clock, DateTimeOffset since)
+        {
+            var currentDateTime = clock.GetCurrentInstant().ToDateTimeOffset();
+            var weekendCompensation = TimeSpan.Zero;
+            var weekendDays = new[] {DayOfWeek.Saturday, DayOfWeek.Sunday};
+
+            for (var date = since.Date; date <= currentDateTime; date += TimeSpan.FromDays(1))
```
Comment (flin-8) [inline source/Cli/Support/ClockExtensions.cs:23]: Instead of compensating, have you considered accumulating instead?

Something like:
```
  if (!weekendDays.Contains(since.Date.DayOfWeek))
                accumulated += TimeSpan.FromDays(1) - since.TimeOfDay;

            for (var date = since.Date; date.Date < currentDateTime.Date; date += TimeSpan.FromDays(1))
            {
                if (!weekendDays.Contains(date.DayOfWeek))
                    accumulated += TimeSpan.FromDays(1);
            }

            if (!weekendDays.Contains(currentDateTime.Date.DayOfWeek))
                accumulated += currentDateTime.TimeOfDay;
```

I think it _might_ make the code easier to reason about? I think it's worth spending an hour on this to see if you can make it cleaner

Comment (sahanocto) [reply source/Cli/Support/ClockExtensions.cs]: I think moving the first and last days outside the for loop won't help too much because of having to check all the partial day scenarios. Instead I went with the accumulating approach so now each day is handled individually and now the return only uses the final sum (removing the `current` - `since` + `compensation`). Also changed it to an early continue for weekends to reduce nesting in the for loop to aid readability.
